### ISSUE TITLE
Convert m2m100 with several languages.

### ISSUE THREAD

**Issue by Jourdelune** [OpenNMT/CTranslate2#689]
Hello, is it possible to convert the m2m100 model with several output languages? Indeed, you have to specify the input and output languages in the converter.

**Comment by guillaumekln** [comment]
Hi,

The source and target languages are optional in the converter. For multilingual models, you should set the `--fixed_dictionary` option instead (same option as in `fairseq-generate`).

Then you check this [forum post](https://forum.opennmt.net/t/convert-m2m-model-to-ctranslate2/4523/6?u=guillaumekln) for an example on how to set the language tokens at runtime.

**Comment by Jourdelune** [comment]
Hi, thanks for your answer!
Do you know what to put in -data_dir?
I get this error: `No such file or directory: '/M2M100/dict.af.txt'`

**Comment by guillaumekln** [comment]
It's the path to the directory containing the dictionaries and other data files. If you downloaded the M2M model from Facebook, you can usually set this path to the model directory you just downloaded.

Here's a command line working on my system:

> `ct2-fairseq-converter --model_path ~/models/facebook/m2m_100/418M_last_checkpoint.pt --data_dir ~/models/facebook/m2m_100/ --fixed_dictionary ~/models/facebook/m2m_100/model_dict.128k.txt --output_dir /tmp/m2m_100_ct2`

(tested with fairseq 0.10.2 and CTranslate2 2.11.0.)

**Comment by Jourdelune** [comment]
Okay, is this [github](https://github.com/pytorch/fairseq/tree/main/examples/m2m_100) that contains m2m100 ?
I have found this url in this github for model_dict.128k.txt.
`wget https://dl.fbaipublicfiles.com/m2m_100/model_dict.128k.txt`

**Comment by guillaumekln** [comment]
Yes, you should head over to the fairseq repository to download this model.

**Comment by Jourdelune** [comment]
Okay, now I just need someone to run the script! (my pc doesn't have the necessary performance)

**Comment by Jourdelune** [comment]
Last question (normally), how to map the vocabulary?
I tested `python build-vmap.py -pt model_dict.128k.txt 1> vmap.txt 2> /dev/null ` without success.

**Comment by guillaumekln** [comment]
This step is not needed to convert and use your model.

You can read in the [README](https://github.com/OpenNMT/CTranslate2#how-do-i-generate-a-vocabulary-mapping-file) what this vocabulary map is and what it is used for. I suggest ignoring this part for now.

**Comment by Jourdelune** [comment]
I run it on google colab only I don't see any acceleration, I take 2,9 seconds to translate with m2m100 the word "Hello world, my name is jourdelune". Is it expected or did I make a mistake somewhere?
`import ctranslate2
import sentencepiece as spm
import time

translator = ctranslate2.Translator("end_output", device="cpu", compute_type="auto")
s = spm.SentencePieceProcessor(model_file='spm.128k.model')

string = "Hello world, my name is jourdelune."
string = s.encode(string, out_type=str)

time1 = time.time()
print("start")
value = translator.translate_batch(
    [string],
    target_prefix=[["__fr__"]],
)
print(value)
print(time.time()-time1)
`

`start
[TranslationResult(hypotheses=[['__fr__', '▁Hello', '▁world', ',', '▁mon', '▁nom', '▁est', '▁jour', 'del', 'une', '.']], scores=[], attention=[])]
2.759963035583496`

**Comment by guillaumekln** [comment]
What is the baseline speed you are comparing to? For example, what's the translation speed when using Fairseq directly?

To take proper measurement, you should also add a warmup phase and average the translation time over multiple samples.

**Comment by Jourdelune** [comment]
I used hugging face and the m2m100 model which takes 3.9 to translate the text (on average, test 50 times). The optimized model takes 2.9 seconds. However in the benchmark with quantization on processor, we see that the translation is multiplied by 5.

**Comment by guillaumekln** [comment]
Can you copy-paste the full code snippet and share it there so that I can test on my system.

**Comment by Jourdelune** [comment]
`!ct2-fairseq-converter --model_path 418M_last_checkpoint.pt --data_dir /M2M100 --quantization int8 --fixed_dictionary model_dict.128k.txt --output_dir end_output  `

```py
import ctranslate2
import sentencepiece as spm
import time

translator = ctranslate2.Translator("/content/drive/MyDrive/end_output/", device="cpu")
s = spm.SentencePieceProcessor(model_file='spm.128k.model')

string = "Hello world, my name is jourdelune."
string = s.encode(string, out_type=str)

value = translator.translate_batch(
    [string],
    target_prefix=[["__fr__"]],
)


print("start")
for i in range(5):
  time1 = time.time()
  value = translator.translate_batch(
      [string],
      target_prefix=[["__fr__"]],
  )
  print(value)
  print(time.time()-time1)```
  
```py
from transformers import M2M100ForConditionalGeneration, M2M100Tokenizer
import time
text = "Hello world, my name is jourdelune."


model = M2M100ForConditionalGeneration.from_pretrained("facebook/m2m100_418M")
tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M")

time1 = time.time()
tokenizer.src_lang = "en"
encoded_hi = tokenizer(text, return_tensors="pt")
generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)

for i in range(5):
  time1 = time.time()
  tokenizer.src_lang = "en"
  encoded_hi = tokenizer(text, return_tensors="pt")
  generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
  print(tokenizer.batch_decode(generated_tokens, skip_special_tokens=True))
  print(time.time()-time1)

```

**Comment by guillaumekln** [comment]
CTranslate2 should be much faster for this example, especially with quantization (verified on my system). Can you try with the following changes:

* Add the missing source language token to the CTranslate2 input:

```python

string = "Hello world, my name is jourdelune."
string = ["__en__"] + s.encode(string, out_type=str)
```

* Set the same number of threads for both frameworks, for example:

```python
import os
os.environ["OMP_NUM_THREADS"] = "2"

import ctranslate2

# rest of the code...
```

**Comment by Jourdelune** [comment]
I went to 1.3. Are there any optimizations still possible to accelerate the model? I make a chat app, so I have to translate texts almost live.

**Comment by Jourdelune** [comment]
I tested on my own pc, and I reach 0,2 seconds, it's great. Is it possible to improve vmap with this model?

**Comment by guillaumekln** [comment]
The vocabulary mapping feature (vmap) can further accelerate the model as demonstrated in the [benchmark](https://github.com/OpenNMT/CTranslate2#benchmarks). You should follow these [instructions](https://github.com/OpenNMT/papers/tree/master/WNMT2018/vmap) to generate the mapping file to be included in the model directory.

Other optimizations include increasing the number of threads (if your hardware allows), reducing the beam size, increasing the batch size when possible, etc.

I'm closing this issue since you managed to convert and run the model. If you have more questions, please consider posting them on the [forum](https://forum.opennmt.net/) instead.
